### PR TITLE
Add support for additional modules in STM32U0.

### PR DESCRIPTION
**Pull request by ExaltZephyr**
This PR adds support for DMA, SPI, RNG, Crypto, USB, RTC, LPUART, LPTIM and WWDG modules in STM32U0.

### CODE REVIEW

**Review by JarmouniA — CHANGES_REQUESTED** [review]

File: dts/arm/st/u0/stm32u0.dtsi
Comment (djiatsaf-st) [inline dts/arm/st/u0/stm32u0.dtsi:318]: When using stm32-dma-v2, I think it is not important to specify `dma-requests` since requests are handled in DMAMUX.

File: dts/arm/st/u0/stm32u0.dtsi
Comment (djiatsaf-st) [inline dts/arm/st/u0/stm32u0.dtsi:330]: ditto

File: dts/arm/st/u0/stm32u0.dtsi
Comment (djiatsaf-st) [inline dts/arm/st/u0/stm32u0.dtsi:384]: According to the refman RM0503, Size of dedicated packet buffer memory SRAM for USB should be `1024` bytes instead of 2048

File: boards/st/nucleo_u083rc/nucleo_u083rc.yaml
Comment (djiatsaf-st) [inline boards/st/nucleo_u083rc/nucleo_u083rc.yaml:23]: Nit  :  sort in alphabetical order with the rest

**Review by djiatsaf-st — CHANGES_REQUESTED** [review]

**Review by FRASTM — CHANGES_REQUESTED** [review]

File: boards/st/nucleo_u083rc/nucleo_u083rc.dts
Comment (ExaltZephyr) [reply boards/st/nucleo_u083rc/nucleo_u083rc.dts:162]: I changed pins because I noticed that PA6 and PA7 conflict with the pins assigned to I2C2 in this file.

File: boards/st/nucleo_u083rc/nucleo_u083rc.dts
Comment (ExaltZephyr) [reply boards/st/nucleo_u083rc/nucleo_u083rc.dts:189]: yes correct, will change pins of spi.

**Review by JarmouniA — CHANGES_REQUESTED** [review]

File: dts/arm/st/u0/stm32u0.dtsi
Comment (djiatsaf-st) [reply dts/arm/st/u0/stm32u0.dtsi:318]: I took a look at your remark, and for me, there will be no issues because if we don't specify, the value 0 is given, which is not bad for this version of dma i think.

Besides, there is a confusion I found only for the `stm32-dma-v2` between `dma-requests` and `dma-channels`. It would have made more sense to me if it was only `dma-channels`. I think that work will be done later to avoid the confusion.

Comment (ExaltZephyr) [reply dts/arm/st/u0/stm32u0.dtsi:318]: I see, but I remember that I had issues when I removed it through testing in past.
will try again to remove it and check what happens

File: dts/arm/st/u0/stm32u0.dtsi
Comment (djiatsaf-st) [inline dts/arm/st/u0/stm32u0.dtsi:327]: to be more compliant , STM32U073xx and STM32U083xx devices have 12 channels and STM32U031xx has 7 channels. 

We will have here  `dma-channels = <7>;`   and in stm32u073.dtsi  we will have at least 
```
dmamux1: dmamux@40020800 
{
	dma-channels = <12>;
	dma-requests = <76>;
};

```

**Review by djiatsaf-st — CHANGES_REQUESTED** [review]

File: dts/arm/st/u0/stm32u0.dtsi
Comment (ExaltZephyr) [reply dts/arm/st/u0/stm32u0.dtsi]: When I removed dma-requests, I encountered issues.
I believe you intended me to remove dma-channels instead of dma-requests, similar to how it's handled on other boards.

I noticed that boards like stm32wb, stm32g0, stm32c0, and stm32wl do not mention dma-channels under the DMA nodes in their DTS configurations. 
To maintain consistency with these boards, I removed dma-channels and added dma-requests instead."

**Comment by ExaltZephyr** [comment]
Hi all,
the twister test "tests/drivers/adc/adc_api" keeps failing. 
During compilation, there's a warning in the adc_stm32 file.
I rebased on main, but the issue is still there. Is there an open bug for this?

**Comment by djiatsaf-st** [comment]
> Hi all, the twister test "tests/drivers/adc/adc_api" keeps failing. During compilation, there's a warning in adc_stm32 file. I rebased on main, but the issue still there. Is there an open bug for this?

Hello @ExaltZephyr , 

Since `nucleo_u083rc` supports DMA (added in `nucleo_u083rc.yaml`), we need to add some configurations in `tests/drivers/adc/adc_api/boards/nucleo_u083rc.overlay` inside `adc1` node to solve the issue.

```
dmas = <&dmamux1 0 5 (STM32_DMA_PERIPH_RX | STM32_DMA_MEM_16BITS |
		STM32_DMA_PERIPH_16BITS)>;
	dma-names = "dmamux";
```

For test bench purposes on our side we use `stm32u083c_dk` instead of `nucleo_u083rc`, I would really appreciate it if you could add this support to `stm32u083c_dk`.

**Review by djiatsaf-st — DISMISSED** [review]

File: boards/st/nucleo_u083rc/doc/index.rst
Comment (erwango) [reply boards/st/nucleo_u083rc/doc/index.rst:203]: @ExaltZephyr Can you comment on this ?

File: drivers/dma/Kconfig.stm32
Comment (erwango) [inline drivers/dma/Kconfig.stm32:59]: Please keep alphabetical order

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/st/nucleo_u083rc/doc/index.rst
Comment (ExaltZephyr) [reply boards/st/nucleo_u083rc/doc/index.rst:203]: @erwango 
I noticed that PA6 and PA7 conflict with the pins assigned to I2C2 in this file,so I made adjustments accordingly.
We now have two options: either retain these pin assignments and disable I2C2, or keep I2C2 enabled and reassign the pins ?

Comment (erwango) [reply boards/st/nucleo_u083rc/doc/index.rst:203]: If possible to have Arduino SPI, Arduino I2C and USB all together, please select this option.
Otherwise please comment I2C and we'll enable it in tests

Comment (ExaltZephyr) [reply boards/st/nucleo_u083rc/doc/index.rst:203]: I2C2 is not present on the Arduino connector, only I2C1 is available. 
Therefore, I will proceed with the second option and disable I2C2.

File: boards/st/nucleo_u083rc/nucleo_u083rc.dts
Comment (erwango) [inline boards/st/nucleo_u083rc/nucleo_u083rc.dts:122]: Not quite clean for the reader.
Let's remove purely the node, if there is no use today, or provide a statement on the reason why it is commented out.

**Review by erwango — CHANGES_REQUESTED** [review]
One very last comment for me. Otherwise, this is fine thanks.

**Review by erwango — APPROVED** [review]

**Comment by ExaltZephyr** [comment]
@erwango 
PR jobs are now passing! 
Is there anything else pending on our end?
Just to note, I'm not authorized to merge.

**Comment by erwango** [comment]
> Is there anything else pending on our end?

No, just waiting more approvals

> Just to note, I'm not authorized to merge.

For sure, me neither. We need to wait to have 2 (green) approvals and then that a release engineer (the ones with merge rights) passes by.

**Review by FRASTM — APPROVED** [review]

**Review by djiatsaf-st — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Pull request merged**
